File: mobile-app/src/services/database/transactionDao.ts
import { v4 as uuidv4 } from 'uuid';
import { getDatabase } from './db';
import { generateTransactionHash } from '@utils/hashEngine';
import type { Transaction, SyncStatus } from '@types/index';

export async function createTransaction(
  params: Omit<Transaction, 'id' | 'integrityHash' | 'syncStatus' | 'createdAt'>,
): Promise<Transaction> {
  const db = getDatabase();
  const id = uuidv4();
  const integrityHash = generateTransactionHash({
    weight: params.weight,
    gpsLat: params.gpsLat,
    gpsLng: params.gpsLng,
    timestampUtc: params.timestampUtc,
    agentId: params.agentId,
  });

  const transaction: Transaction = {
    ...params,
    id,
    integrityHash,
    syncStatus: 'pending',
    createdAt: new Date().toISOString(),
  };

  await db.executeSql(
    `INSERT INTO transactions
      (id, commodity_type, weight, unit, buyer_id, seller_id, gps_lat, gps_lng,
       gps_accuracy, timestamp_utc, integrity_hash, sync_status, agent_id, notes, created_at)
     VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`,
    [
      transaction.id, transaction.commodityType, transaction.weight, transaction.unit,
      transaction.buyerId, transaction.sellerId, transaction.gpsLat, transaction.gpsLng,
      transaction.gpsAccuracy, transaction.timestampUtc, transaction.integrityHash,
      transaction.syncStatus, transaction.agentId, transaction.notes ?? null,
      transaction.createdAt,
    ],
  );

  // Add to sync queue
  await db.executeSql(
    `INSERT INTO sync_queue (transaction_id, status) VALUES (?, 'pending')`,
    [id],
  );

  return transaction;
}

export async function getTransactionById(id: string): Promise<Transaction | null> {
  const db = getDatabase();
  const [result] = await db.executeSql(
    'SELECT * FROM transactions WHERE id = ?',
    [id],
  );
  if (result.rows.length === 0) return null;
  return mapRow(result.rows.item(0));
}

export async function listTransactions(agentId: string): Promise<Transaction[]> {
  const db = getDatabase();
  const [result] = await db.executeSql(
    'SELECT * FROM transactions WHERE agent_id = ? ORDER BY created_at DESC',
    [agentId],
  );
  return Array.from({ length: result.rows.length }, (_, i) => mapRow(result.rows.item(i)));
}

export async function updateTransactionSyncStatus(
  id: string,
  status: SyncStatus,
): Promise<void> {
  const db = getDatabase();
  await db.executeSql(
    'UPDATE transactions SET sync_status = ? WHERE id = ?',
    [status, id],
  );
}

export async function getPendingTransactions(): Promise<Transaction[]> {
  const db = getDatabase();
  const [result] = await db.executeSql(
    "SELECT * FROM transactions WHERE sync_status = 'pending' ORDER BY created_at ASC",
  );
  return Array.from({ length: result.rows.length }, (_, i) => mapRow(result.rows.item(i)));
}

function mapRow(row: Record<string, unknown>): Transaction {
  return {
    id: row.id as string,
    commodityType: row.commodity_type as string,
    weight: row.weight as number,
    unit: row.unit as string,
    buyerId: row.buyer_id as string,
    sellerId: row.seller_id as string,
    gpsLat: row.gps_lat as number,
    gpsLng: row.gps_lng as number,
    gpsAccuracy: row.gps_accuracy as number,
    timestampUtc: row.timestamp_utc as string,
    integrityHash: row.integrity_hash as string,
    syncStatus: row.sync_status as SyncStatus,
    agentId: row.agent_id as string,
    notes: row.notes as string | undefined,
    createdAt: row.created_at as string,
  };
}

File: mobile-app/lib/core/database/transaction_dao.dart
import 'package:sqflite_sqlcipher/sqflite.dart';
import 'package:uuid/uuid.dart';
import '../crypto/hash_engine.dart';
import 'database_service.dart';

const _uuid = Uuid();

class TransactionDao {
  final Database _db;
  TransactionDao(this._db);

  factory TransactionDao.fromSingleton() => TransactionDao(getDatabase());

  Future<String> insert({
    required String commodityType,
    required double weight,
    required String unit,
    required String buyerId,
    required String sellerId,
    required double gpsLat,
    required double gpsLng,
    double? gpsAccuracy,
    required String agentId,
    String? notes,
  }) async {
    final id = _uuid.v4();
    final timestamp = DateTime.now().toUtc().toIso8601String();
    final hash = generateTransactionHash(
      weight: weight,
      gpsLat: gpsLat,
      gpsLng: gpsLng,
      timestampUtc: timestamp,
      agentId: agentId,
    );

    await _db.insert('transactions', {
      'id': id,
      'commodity_type': commodityType,
      'weight': weight,
      'unit': unit,
      'buyer_id': buyerId,
      'seller_id': sellerId,
      'gps_lat': gpsLat,
      'gps_lng': gpsLng,
      'gps_accuracy': gpsAccuracy,
      'timestamp_utc': timestamp,
      'integrity_hash': hash,
      'sync_status': 'pending',
      'agent_id': agentId,
      'notes': notes,
    });

    await _db.insert('sync_queue', {
      'transaction_id': id,
      'retry_count': 0,
      'status': 'pending',
    }, conflictAlgorithm: ConflictAlgorithm.ignore);

    return id;
  }

  Future<List<Map<String, dynamic>>> getPending() => _db.query(
        'transactions',
        where: 'sync_status = ?',
        whereArgs: ['pending'],
        orderBy: 'created_at DESC',
      );

  Future<List<Map<String, dynamic>>> getAll({String? agentId}) {
    if (agentId != null) {
      return _db.query(
        'transactions',
        where: 'agent_id = ?',
        whereArgs: [agentId],
        orderBy: 'created_at DESC',
      );
    }
    return _db.query('transactions', orderBy: 'created_at DESC');
  }

  Future<Map<String, dynamic>?> getById(String id) async {
    final rows = await _db.query('transactions', where: 'id = ?', whereArgs: [id]);
    return rows.isEmpty ? null : rows.first;
  }

  Future<void> markSynced(String id) => _db.update(
        'transactions',
        {'sync_status': 'synced'},
        where: 'id = ? AND sync_status != ?',
        whereArgs: [id, 'synced'],
      );

  Future<void> markFailed(String id) => _db.update(
        'transactions',
        {'sync_status': 'failed'},
        where: 'id = ?',
        whereArgs: [id],
      );
}
